File: leet_0075_sort_colors.cc
class Solution {
public:
    void sortColors(vector<int>& nums)
    {
        int len = nums.size() ;

        if (len < 2)
            return ;
        if (len == 2)
        {
            if (nums[0] > nums[1])
                swap(nums[0], nums[1]) ;
            return ;
        }

        int i = 0, j = len - 1, k ;

        while (!nums[i])
            ++ i ;
        while (nums[j] == 2)
            -- j ;
        k = i - 1;
        while (++k <= j)
        {
            if (!nums[k])
            {
                nums[k] = nums[i];
                nums[i++] = 0;
            }
            else if (nums[k] == 2)
            {
                nums[k] = nums[j];
                nums[j--] = 2;
                k--;
            }
        }
    }
};

File: leet_0075_sort_colors.scala
object Solution {

    import scala.collection.mutable.ArrayBuffer // .swap

    def sortColors(nums: Array[Int]): Unit = {

        val N = nums.length
        var l = 0
        var r = N - 1
        while (l < N && nums(l) == 0) {
            l += 1
        }
        while (r > -1 && nums(r) == 2) {
            r -= 1
        }
        var mid = l
        while (mid < r + 1) {
            if (nums(mid) == 0) {
                val temp = nums(mid)
                nums(mid) = nums(l)
                nums(l) = temp
                l += 1
            } else if (nums(mid) == 2) {
                val temp = nums(mid)
                nums(mid) = nums(r)
                nums(r) = temp
                r -= 1
                mid -= 1
            }
            mid += 1
        }
    }
}
